File: c/0912-sort-an-array.c
void merge(int *array,int l,int m, int r){

    int n1 = m-l+1;
    int n2 = r-m;

    int * left = (int *)malloc(sizeof(int)*n1);
    int * right=(int *)malloc(sizeof(int)*n2);

    for(int i=0;i<n1;i++){
        left[i]=array[l+i];
    }

    for(int j=0;j<n2;j++){
        right[j]=array[m+j+1];
    }

    int i=0;
    int j=0;
    int k=l;

    while(i<n1 && j<n2){
        if(left[i]<=right[j]){
            array[k]=left[i];
            i++;
        }else{

            array[k]=right[j];
            j++;
        }
        k++;
    }

    while(i<n1){
        array[k]=left[i];
        i++;
        k++;
    }

    while(j<n2){
        array[k]=right[j];
        j++;
        k++;
    }

}

void mergesort(int * array, int l,int r){

    if(l<r){
        int m = (l+r)/2;

        mergesort(array,l,m);
        mergesort(array,m+1,r);

        merge(array,l,m,r);
    }

}

int* sortArray(int* nums, int numsSize, int* returnSize) {

    int *array = (int *) malloc(sizeof(int)*numsSize);
    for(int i=0;i<numsSize;i++){

        array[i]=nums[i];
    }

    mergesort(array,0,numsSize-1);

    *returnSize=numsSize;

    return array;
    
}

File: swift/0912-sort-an-array.swift
/**
 * Question Link: https://leetcode.com/problems/sort-an-array/
 */

 class Solution {
    func sortArray(_ nums: [Int]) -> [Int] {
        var nums = nums

        func merge(arr: inout [Int], l: Int, m: Int, r: Int) {
            var left = Array(arr[l...m])
            var right = Array(arr[m + 1...r])
            var i = l
            var j = 0
            var k = 0
            
            while j < left.count && k < right.count {
                if left[j] <= right[k] {
                    arr[i] = left[j]
                    j += 1
                } else {
                    arr[i] = right[k]
                    k += 1
                } 
                i += 1
            }

            while j < left.count {
                arr[i] = left[j]
                j += 1
                i += 1
            }

            while k < right.count {
                arr[i] = right[k]
                k += 1
                i += 1
            }

        }
        
        func mergeSort(arr: inout [Int], l: Int, r: Int) -> [Int] {
            if l == r {
                return arr
            }

            let m = (l + r) / 2
            mergeSort(arr: &arr, l: l, r: m)
            mergeSort(arr: &arr, l: m + 1, r: r)
            merge(arr: &arr, l: l, m: m, r: r)
            return arr
        }

        return mergeSort(arr: &nums, l: 0, r: nums.count - 1)
    }
}